go/pkg/client/tree.go

```diff
@@ -523,16 +517,17 @@ func packageDirectories(t *treeNode) (root *repb.Directory, files map[digest.Dig
 	}
 	sort.Slice(root.Directories, func(i, j int) bool { return root.Directories[i].Name < root.Directories[j].Name })
 
-	for name, fn := range t.files {
-		dg := fn.ue.Digest
-		root.Files = append(root.Files, &repb.FileNode{Name: name, Digest: dg.ToProto(), IsExecutable: fn.isExecutable})
-		files[dg] = fn.ue
+	for name, n := range t.leaves {
+		if n.file != nil {
+			dg := n.file.ue.Digest
+			root.Files = append(root.Files, &repb.FileNode{Name: name, Digest: dg.ToProto(), IsExecutable: n.file.isExecutable})
+			files[dg] = n.file.ue
+		}
+		if n.symlink != nil {
```

Comment:
Same comment here about precedence and unexpected nodes.

Reply:
Added a consistency check. I think erroring out is better here, because this should never happen (and if it does, we have a bug somewhere in this file, because it's an internal type). Would have been easier to write in a language where we could enforce immutability and constructors (because then we could just make sure that all fileSysNodes are created with exactly one of these variables set).

Reply:
Hmm, I'm having a second thought since this is a internal error that the user has no way of handling. I'm leaning towards maintaining existing semantics by just letting a file node take precedence. Thoughts?

Reply:
Note that the current code maintains the existing semantics precisely -- we used to keep the file and symlink nodes in separate arrays and go over all of them independently. So there was no "precedence" -- we would go over all of them, and if there was an inconsistency, we would encode both the file and the symlink (creating an invalid tree proto).

I think: either we want to check internal consistency or we trust our current code and don't check it. I'm leaning towards checking/enforcing, since "trust our current code" is famous last words :-)
Then, the only question is how to make the existing code more readable (to make it clear that a fileSysNode is a union). How about this:
- add a `type` field in `fileSysNode` which is one of:

```
type NodeType int

const (
    File NodeType = iota
    Symlink
    Directory
)
```

And then instead of the if cases, do a `switch` over the `type` field. This would change the existing semantics, but it would be a bit clearer what's going on here. WDYT?

Reply:
I thought the file node was taking precedence over the symlink in this if-else statement: https://github.com/bazelbuild/remote-apis-sdks/pull/466/files#diff-45c6218ebcf3e4b4c948697c226c2d8d9ef4a3f8b1e22d420cb4dc655e74e143L350

Isn't the marker field just a more verbose way of preferring a field over another? It's still possible to have multiple fields set, but now the node type field determines which one is actually used.

I think a robust solution would need to involve generics or an interface-based polymorphism. Both of which sound out of scope for your purposes.

I'm happy to approve a compromise where we maintain existing semantics.

Reply:
I think just adding a `continue` above this line would do it.

Reply:
> I thought the file node was taking precedence over the symlink in this if-else statement:

Yes, but then we trust the arrays under the treeNode to be correct in a different function. The code is internally consistent and fileSysNode is always created and used as a union. Yeah, maybe since it's internal we don't want to enforce it with a user-visible error...

> I think a robust solution would need to involve generics or an interface-based polymorphism.

Yeah, there is no truly robust solution in Go (need to enforce immutability and constructor usage), and trying to add one would go against the grain of the language, I think (Go is, like, "it's okay, we all know what we're doing"...)

Sure, I added an else case and a comment instead, PTAL.

Reply:
Sounds good.

One last nit if you don't mind: [`continue` rather than `else`](https://google.github.io/styleguide/go/decisions#indent-error-flow)

Reply:
Done, although here I disagree:
1. This is not error flow (so the style guideline does not apply).
2. The indent is the same either way (so continue will not make the rest of the code easier to read).
3. `else` makes it a clearer, imo, that the cases are intended as disjoint. This is semantically a `switch` statement.

Up to you.

Thank you!

Reply:
Thanks.

I think this is more consistent with the style guide, which cannot be comprehensive. Code readability is highly subjective and I try to keep it as close to the style guide as possible, even when I disagree myself, which helps with consistency across code bases.